### ISSUE TITLE
Speed up of the LES page by using ajax request to only request particular subsets of the data

### ISSUE THREAD

**Issue by grdw** [quintel/etmoses#703]
Right now all the `35040` data points for all end-points are returned to the front-end. Resulting in some slowness in the front-end, particularly for some really larges LES's.

The full calculation is first calculated, than cached. The `TestingGround::Calculator` should be able to take three extra arguments. A `range`, a `resolution` and a `node` to take specific parts from the cache or write a new class which would be able to that? (@antw  do you have an opinion about this?).

When initially visiting a LES the full calculation should execute. When you click on a node another request should fetch the low-resolution image with [downsampleCurve](https://github.com/quintel/etmoses/blob/master/app/assets/javascripts/testing_grounds.js.coffee#L174-L184) for that node. The downsampling should be moved to the Ruby code. 

When changing the date (with the select box at the bottom of the graph), it should do another ajax request with a specified range and query the data from cache.

Some pseudo-code:
````ruby
class TestingGround
  class CacheFetcher
    def initialize(testing_ground, strategies)
       @testing_ground = testing_ground
       @strategies = strategies
    end
    
    def fetch(node, resolution = :high, range = 0..35040)
      NetworkCache::Fetcher.from(@testing_ground, @strategies).
        fetch_node(node).
        with_resolution(resolution).
        within_range(range)
    end
  end
end
````

**Comment by ChaelKruip** [comment]
@grdw glad that you are thinking about this! next year, we'll have budget to work on this and other optimisation topics. Could you make a proper design for how this should work and include @antw in the choices you make?

**Comment by grdw** [comment]
I have a technical design of it [here](https://gist.github.com/grdw/bacfbfb46f5887424938). Maybe you have some thoughts on this @antw.

**Comment by antw** [comment]
I like the idea to make the topology and load data JSON separate.

--

> An ajax request fires to fetch a node and the corresponding load in a low resolution.

I think we will have to fetch the low-res data for *all* nodes once the calculation is finished; this is how we know which (over-capacity) nodes to highlight with the red border.

--

> When somebody clicks a node [...] `Ajax.json(url, { node: 'HV' },`

Instead of `{ node: '<name>' }`, can I suggest instead accepting an *array* of node names? For example:

```js
Ajax.json(url, { nodes: ['HV', 'MV'] },  function() {}, function() {});
```

In my mind I was thinking something like this:

```
1. Get topology JSON

   GET /topologies/:topology_id.json

2. Extract all the node names from the topology and request the loads (triggers the
   calculation if necessary).

   GET /testing_grounds/:id/data.json?nodes=HV,MV&resolution=low

3. User clicks a node.

   (no request, low res data has already been loaded)

4. User changes resolution to show one week. Fetch the high-res data.

   GET /testing_grounds/:id/data.json?nodes=<node_name>&resolution=high
```

But don't take this as gospel! Feel free to experiment with other ideas.

--

> When changing the date [...] `range: [0, 1000]`

I like the thinking here, but for a first implementation I would like to omit this feature. It adds complexity, and I'm not convinced it's necessary.

Pros:
* The first time the user views high-resolution data for a node, that data will load faster.

Cons
* An extra HTTP request each time the user changes the week ("laggy" experience for users when changing weeks).
* Added complexity in JS and Ruby.

Can we try *without* the `range` feature initially, and see how it performs?

As an aside, if we do implement it later, I would like to suggest that the `range` takes *dates* instead of numbers: they're human-readable *and* more future-proof. For example if, in the future, we add testing grounds with a resolution of 60 minutes – or 5 minutes – instead of 15, a date is compatible with both, but a numeric range is not:

| Request                    | 15 min res      | 60 minute res  |
|----------------------------|-----------------|----------------|
| `range: ['Jan 1', 'Jan 7]` | Jan 1 - Jan 7   | Jan 1 - Jan 7  |
| `range: [0, 672]`          | Jan 1 - Jan 7   | Jan 1 - Jan 28 |
| `range: [34367, 35039]`    | Dec 24 - Dec 31 | Invalid        |

--

> When there's a node (and range) present in `localStorage`

I'm skeptical about using localStorage. There are limits to how much you can store (2.5MB - 10MB depending on the browser) and it burdens you with having to manage this client-side cache. For a first implementation, how about just storing the loads in a JS object inside a `TestingGround` class?

*Very* rough idea:

```js
var les = new TestingGround(id);

// Fetches the data from the server.
les.nodeLoad(['HV', 'MV'], { resolution: 'low' })
  // Callback when data is retrieved.
  .done(function() {})
  // Callback when calculation failed.
  .fail(function() {}};

// Requesting the data again (for some reason). Data has been requested previously, and
// was stored inside "les". The "done" callback runs immediately.
les.nodeLoad(['HV', 'MV'], { resolution: 'low' })
  .done(function() {})
  .fail(function() {}};

// User wants high resolution data for the HV node. We don't have this yet, so a request
// is made.
les.nodeLoad(['HV'], { resolution: 'high' })
  .done(function() {})
  .fail(function() {}};
```

I can go into more detail on this if you'd like?

--

> 1. Calculating the network partially, per node
> Would this be possible or would this (when inspecting 4 nodes for instance) be slower?

Not possible. :disappointed: 

The network has to be calculated as a whole because the loads of node A may affect the loads of node B. As a practical example: Node A has a very large base load and causes the MV node to become over-capacity. Node B therefore postpones some of it's base load.

If we only compute node B, it won't react properly to the congestion. Therefore, to get accurate behaviour from the technologies, the whole network has to be calculated.

**Comment by grdw** [comment]
@antw thanks for your remarks.

I agree with the fact that the resolution `low` and `high` should be incorporated first. The `range` can be done later. I'm also not sure how much the range will impact the speed (mainly because the whole cache needs to be read anyway).  It can take away some slowness if you selected a range and than reload the page.

> I'm skeptical about using localStorage. There are limits to how much you can store

The idea was not to store the loads in `localStorage`, I'm sorry that that wasn't really clear. But basically only the `range` and the `resolution`. The `range` is alredy stored in `localStorage`. 

> The network has to be calculated as a whole because the loads of node A may affect the loads of node B. 

I already thought so. Doesn't matter though.

**Comment by grdw** [comment]
> I think we will have to fetch the low-res data for all nodes once the calculation is finished; this is how we know which (over-capacity) nodes to highlight with the red border.

Credits to @ChaelKruip. We could also do it in this order:
````
1. Get topology JSON

   GET /testing_grounds/:id/data.json?nodes=&resolution=low

   And in this case fetch a topology with no loads but with an `over_capacity` flag to draw the nodes red.

2. Extract all the node names from the topology and request the loads (triggers the
   calculation if necessary).

   GET /testing_grounds/:id/data.json?nodes=HV,MV&resolution=low

3. User clicks a node.

   (no request, low res data has already been loaded)

4. User changes resolution to show one week. Fetch the high-res data.

   GET /testing_grounds/:id/data.json?nodes=<node_name>&resolution=high

````

**Comment by antw** [comment]
> The idea was not to store the loads in localStorage. But basically only the range and the resolution properties. The range is already stored in localStorage.

Got it; this is so that selecting a new node remembers the chart settings from the previous one?

> I'm also not sure how much the range will impact the speed (mainly because the whole cache needs to be read anyway)

Good point. :+1:

> And in this case fetch a topology with no loads but with an `over_capacity` flag to draw the nodes red.

The only difference I see is an `over_capacity` flag with the topology (perhaps I'm missing something? :laughing:). This moves the burden for detecting over-capacity nodes onto the server. If we're going to read the low-res data out of the cache to do this, is it *that* much more expensive to send it to the browser at the same time? Maybe yes, maybe no! Worth testing.

> *Very* rough idea. [...] I can go into more detail on this if you'd like?

To explain a bit more: try to think in high-level classes as much as possible. Of these two code snippets, the second is probably much easier for an ETMoses newcomer to understand.

```js
Ajax.json("http://moses.dev/testing_grounds/" + id + "/data.json",
  { nodes: ['HV'], strategies: { capping_solar_pv: true, ... } }, 
  function(data, response, xhr) {}, 
  function(error, xhr){}
);
```

```js
var strategies = new Strategies({ capping_solar_pv: true, ... }),
    les        = new TestingGround(id, strategies);

// Fetches the data from the server.
les.nodeLoad(['HV', 'MV'], { resolution: 'low' })
  .done(function(loads) {})
  .fail(function() {}};
```

The first translates as: send a JSON HTTP request to *this* address with *that* JSON payload, and with the response do a *thing*.

The second translates as: With *this* LES, give me the loads for *these nodes*, and then do *something* with them.

--

Here's a cool side-effect of TestingGround#nodeLoad returning a jQuery deferred: we could use [jQuery.when](https://api.jquery.com/jquery.when/) to do the "basic" and "features" requests simultaneously, and then perform some act when both requests are complete:

```js
var strategies  = new Strategies({ capping_solar_pv: true, ... }),
    basicLES    = new TestingGround(id)
    featuresLES = new TestingGround(id, strategies),

    nodes       = ['HV', 'MV'],
    options     = { resolution: 'low' };

// Fetches the data from the server.
jQuery.when(basicLES.nodeLoad(nodes, options), featuresLES.nodeLoad(nodes, options))
  // Both requests completed:
  .done(function(basicLoads, featuresLoads) {})
  // One or both requests failed:
  .fail(function() {}};
```

**Comment by grdw** [comment]
> Here's a cool side-effect of TestingGround#nodeLoad returning a jQuery deferred: we could use jQuery.when to do the "basic" and "features" requests simultaneously,

I know of `jQuery.when` :+1:  good suggestion though.

**Comment by grdw** [comment]
Current status:

- [x] Rewrote the current javascript in the form of:

````javascript
[ new Les(), new Les(strategies) ];
````

- [x] Wrote a tiny ruby class that downsamples the current graph to 365 points

Todo:
- [ ] Change strategies should redraw correctly (still testing)
- [ ] When changing the date it should change resolution for the current chart and all previously clicked loads
